### PR TITLE
Consolidate eBPF redirection

### PR DESCRIPTION
**Pull request by PlatformLC**
**Please provide a description of this PR:**

Consolidate eBPF redirection behavior as desired:

1. As long as Pod should be enrolled in ambient, configure the redirection rules and intercept the traffic from very beginning stage (cni-plugin point) no matter if ztunnel is ready
2. Reconcile Pods IP to avoid map leak in case some Pods were deleted during the outage period of isito-cni
3. Drop packets if ztunnel is unavailable and remain eBPF redirection rule while istio-cni existed

But we need some trigger or hook point(istioctl, operator or helm?) to cleanup the redirection rules for case istio is truly uninstalled from the system.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [x] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by PlatformLC** [comment]
@bleggett @howardjohn PTAL for this change？

File: cni/pkg/ambient/net_linux.go [hunk 1]
```diff
@@ -1189,9 +1190,6 @@ func (s *Server) CreateRulesWithinNodeProxyNS(proxyNsVethIdx int, ztunnelIP, ztu
 func (s *Server) cleanupNode() {
 	log.Infof("Node-level network rule cleanup started")
 	if s.redirectMode == EbpfMode {
-		if err := s.delZtunnelEbpfOnNode(); err != nil {
```
Comment (howardjohn) [inline cni/pkg/ambient/net_linux.go:1192]: why do we skip cleanup on ebpf but not iptables? I would think the logic should be consistent?

File: cni/pkg/ebpf/server/redirectionServer_linux.go [hunk 2]
```diff
@@ -581,19 +582,16 @@ func (r *RedirectServer) dumpZtunnelInfo() (*mapInfo, error) {
...
 	var valueOut mapInfo
-	var values []mapInfo
-	var keys []uint32
+	m := sets.New[string]()
 	mapIter := r.obj.AppInfo.Iterate()
 	for mapIter.Next(&keyOut, &valueOut) {
-		keys = append(keys, keyOut)
-		values = append(values, valueOut)
-
+		ipAddr := netip.AddrFrom4(keyOut)
```
Comment (howardjohn) [inline cni/pkg/ebpf/server/redirectionServer_linux.go:591]: todo: support v6?

File: cni/pkg/ambient/net_linux.go (hunk 1 above)
Comment (irisdingbj) [reply cni/pkg/ambient/net_linux.go:1192]: This is to make sure traffic will still go through ztunnel if istio-cni is broken/stopped.   Without this, if istio-cni is stopped working, the existing traffic to ztunnel will be bypassed as well.

File: cni/pkg/plugin/plugin.go [hunk 3]
```diff
@@ -220,7 +220,7 @@ func CmdAdd(args *skel.CmdArgs) (err error) {
 
 			log.Debugf("ambientConf.ZTunnelReady: %v", ambientConf.ZTunnelReady)
 			added := false
-			if !excludePod && ambientConf.ZTunnelReady {
```
[the remark this answers is not in the document]
Comment (irisdingbj) [reply cni/pkg/plugin/plugin.go:223]: If ztunnel is crashing, this change will drop the packet directly : https://github.com/istio/istio/pull/44208/files#diff-c3b29496fc3e6f9e25813ee08e098497adfa7061d3d3e24582892c39996183f8R135  What do you mean ' keep forwarding`?

File: cni/pkg/ambient/server.go [hunk 4]
```diff
@@ -245,11 +252,16 @@ func (s *Server) ReconcileZtunnel() error {
...
 	}
-
+	existed := s.getEnrolledIPSets()
 	// Reconcile namespaces, as it is possible for the original reconciliation to have failed, and a
 	// small pod to have started up before ztunnel is running... so we need to go back and make sure we
 	// catch the existing pods
-	s.ReconcileNamespaces()
+	processed := s.ReconcileNamespaces()
+
+	stales := existed.Difference(processed)
```
[the remark this answers is not in the document]
Comment (PlatformLC) [reply cni/pkg/ambient/server.go:261]: Per Difference implementation, current logic should be ok.

// Difference returns a set of objects that are not in s2
// For example:
// s = {a1, a2, a3}
// s2 = {a1, a2, a4, a5}
// s.Difference(s2) = {a3}
// s2.Difference(s) = {a4, a5}

File: cni/pkg/ebpf/server/redirectionServer_linux.go (hunk 2 above)
Comment (PlatformLC) [reply cni/pkg/ebpf/server/redirectionServer_linux.go:591]: yes, v6 support is a todo list.

File: cni/pkg/ambient/server.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (irisdingbj) [reply cni/pkg/ambient/server.go:261]: yep, you are right

File: cni/pkg/ambient/net_linux.go (hunk 1 above)
Comment (howardjohn) [reply cni/pkg/ambient/net_linux.go:1192]: There is also (3) CNI is stopped because its being replaced. We should NOT clean up here

File: cni/pkg/plugin/plugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (howardjohn) [reply cni/pkg/plugin/plugin.go:223]: What I mean by "latching" is once we program redirection we should not turn it off unless the user took an explicit action to uninstall.

0. Nothing installed
1. Ztunnel, CNI deployed but not yet `Ready` - we should have no rules
2. Ztunnel becomes ready - we should program redirection
3. Ztunnel becomes unready - we should NOT undo redirection

File: cni/pkg/ambient/net_linux.go (hunk 1 above)
Comment (bleggett) [reply cni/pkg/ambient/net_linux.go:1192]: @howardjohn 👍  - we have no mechanism to determine (3) yet tho - that's what you mean by `latching` I guess.

Comment (howardjohn) [reply cni/pkg/ambient/net_linux.go:1192]: Yep exactly. Strawman proposal would be just to write to the host disk (we already do this for some other settings). but we may need ot think it through more

File: cni/pkg/plugin/plugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bleggett) [reply cni/pkg/plugin/plugin.go:223]: I guess I'm still not convinced we really need to care about zt readiness as a redirection rule gate tho. It's simpler to just let CNI do what it needs to do as soon as it knows it will need to.

That way _as soon as_ the user indicates to the control plane that redirection is desired we start redirecting. If ztunnel isn't ready to proxy then the traffic will be dropped, but that seems less surprising than not redirecting at all until ztunnel is ready, whenever that happens to be.

The main benefit of tying CNI readiness and ztunnel readiness together into a single state is it allows for e.g. node scheduling taints as mentioned here: https://github.com/istio/istio/issues/40303 to work more simply, I guess.

[the remark this answers is not in the document]
Comment (howardjohn) [reply cni/pkg/plugin/plugin.go:223]: My concern with that is if you just do `istioctl install` then the redirect starts happening before ztunnel is deployed and the install is not zero-downtime.

[the remark this answers is not in the document]
Comment (bleggett) [reply cni/pkg/plugin/plugin.go:223]: That at least is something I would be OK fixing with an out of band "wait for ztunnels" `istioctl install` gate, like we do already for ingressgateways/CNI/etc.

[the remark this answers is not in the document]
Comment (howardjohn) [reply cni/pkg/plugin/plugin.go:223]: I think helm has a --wait and we can specify to install ztunnel first probably.

I am not sure this solves "new nodes" issues. We might have other issues for new nodes though, since we lose the ability to use our workaround for the race in https://github.com/istio/istio/issues/40303

File: cni/pkg/ambient/net_linux.go (hunk 1 above)
Comment (PlatformLC) [reply cni/pkg/ambient/net_linux.go:1192]: Let me have an summary for _"stopping"_ case we discussed above:
1. CNI _crashes_ or terminates abnormally  --> everything may be indeterminate, new cni POD should restore the desired state
2. CNI is _graceful stopped_ due to _uninstall_   --> we'd like to do cleanup
3. CNI is _graceful stopped_ due to _upgrade/replacement_   --> we'd like to latch rules  to make sure no escaping during upgrade

So question comes to how to identify it's _uninstall_ or _upgrade/replacement_. Since there's no mechanism, I think we could move the _cleanup_ stuff  into `istioctl` or `helm`  `pre-delete` job and do nothing for the stop procedure of istio-cni process.

File: cni/pkg/ambient/net_linux.go [hunk 5]
```diff
@@ -1208,6 +1206,23 @@ func (s *Server) cleanupNode() {
 	}
 }
 
+func (s *Server) getEnrolledIPSets() sets.Set[string] {
```
[the remark this answers is not in the document]
Comment (PlatformLC) [reply cni/pkg/ambient/net_linux.go:1209]: Hmm, I think ipsets is not only used for DNS redir, but also for tcp outbound of Pods, refer to [rules](https://github.com/istio/istio/blob/ce1cef0a9866ed40c746c7e1988be479cb207ef8/cni/pkg/ambient/net_linux.go#L576)

File: cni/pkg/plugin/plugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (PlatformLC) [reply cni/pkg/plugin/plugin.go:223]: I'm not very sure how to understand 

> we lose the ability to use our workaround for the race in https://github.com/istio/istio/issues/40303

I think https://github.com/istio/istio/issues/40303 is to guarantee `istio-cni` is ready before pods being scheduled on that node?

1. New node added
2. istio-cni is ready
3. pods scheduled on this node by k8s

The sequence is guaranteed by taint controller. With this, `istio-cni` could catch each pods event and make sure all traffic could be intercepted ASAP without any exception 

Think further, without https://github.com/istio/istio/issues/40303, (2) and (3) would be reversed. Thus, before `istio-cni` ready, traffic will escape until `ztunnel` is ready.

So I think the main concern is whether we need do a immediate reconciliation while `istio-cni` startup rather than being triggered by `ztunnel` pod event?

File: cni/pkg/ambient/net_linux.go (hunk 1 above)
Comment (bleggett) [reply cni/pkg/ambient/net_linux.go:1192]: >So question comes to how to identify it's uninstall or upgrade/replacement. Since there's no mechanism, I think we could move the cleanup stuff into istioctl or helm pre-delete or post-delete job and do nothing for the stop procedure of istio-cni process.

The problem is that the cleanup logic is tightly tied to the creation logic, and like the creation logic is complex/specific/requires a lot of the same support tooling.

Cleanup and teardown should live in the same spot. Separating them seems like a source of bad complexity. Rule creation cannot live "outside" of the CNI watch server, by definition. The problem is the fact that our CNI watchserver manages itself and is unaware of upgrades.

I kinda don't like Helm lifecycle hooks, as they tend to open up a Pandora's box of manually overriding resource creation order, manually garbage-collecting hook-created resources, etc etc. But if we are careful this is an option. The other problem with this is that as mentioned above we would essentially be duplicating much of the rule creation code in the `teardown` hook init/teardown container binary.

In much the same way that I am suspicious of Helm lifecycle hooks, I am suspicious of bespoke CRDs, but some of what Costin talked about https://github.com/istio/istio/issues/43642 WRT dropping the CNI plugin model  and just creating a CRD that manages rules and upgrades begins to make sense here.

File: cni/pkg/plugin/plugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bleggett) [reply cni/pkg/plugin/plugin.go:223]: > I think #40303 is to guarantee `istio-cni` is ready before pods being scheduled on that node?

Yes - waiting for `istio-cni` readiness as mentioned in #40303 isn't enough anymore if we want to prevent pods from being scheduled before redirection AND proxying are ready - since gating on `istio-cni` readiness doesn't guarantee the latter at all (or even really the former - we might need to check both in all cases).


## Scenario 1: `istio-cni` "holds/waits" (via node taint or whatever else) until `ztunnel` is also ready on that node:

- To guarantee no traffic escapes -> we need to wait for `istio-cni` readiness (which implicitly also waits for ztunnel readiness).
- To guarantee pods are not scheduled until traffic _can be successfully proxied_ -> same as above.

 ## Scenario 2: `istio-cni` does NOT "hold/wait" (via node taint or whatever else) until `ztunnel` is also ready on that node:

- To guarantee no traffic escapes -> we need to wait for `istio-cni` readiness on that node.
- To guarantee pods are not scheduled until traffic _can be successfully proxied_ -> we need to wait for `ztunnel` readiness on that node, separately via some other mechanism.

File: cni/pkg/ambient/net_linux.go (hunk 1 above)
Comment (PlatformLC) [reply cni/pkg/ambient/net_linux.go:1192]: If we could find a good mechanism to identify it's an uninstall or upgrade/crash/terminating event with the daemonset itself,  we should cleanup everything at the same context created it.

Regarding to CRD, IMHO, I think the main idea of it is also to have the high dimension view of what exact task we need to do, it's more like a 'scavenger' operator who is the first and last one of all.

File: cni/pkg/plugin/plugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (PlatformLC) [reply cni/pkg/plugin/plugin.go:223]: There're 2 points we're talking and thinking about: 1. Traffic escaping. 2. Traffic READY
Node taint is more meaningful for preventing traffic escaping, but traffic READY is based on AND of both cni and ztunnel.

So theoretically if we guarantee Pods scheduled after cni ready, we could make sure no escaping, but for some scenario like cni upgrade crash, it's still possible we have a gap interval (with plugin removed and no watcher), so reconciliation at cni startup looks more reasonable.

It looks OOS for the original change.

The original change is to reduce the possibility of traffic escaping. I think it's not harmful and let's move forward.

[the remark this answers is not in the document]
Comment (bleggett) [reply cni/pkg/plugin/plugin.go:223]: Removing `s.delZtunnelEbpfOnNode()` from `nodeCleanup` ~is~ sort of harmful, though - right? That's my main problem with this PR.

- it means prog redirection is left in place in a cluster on uninstall (not current behavior) leaving clusters in a broken state on uninstall.

- it means prog redirection is left in place on uninstall (not current behavior) meaning if you reinstall to the same cluster and switch to `iptables` (a common testing scenario) network redirection will be badly broken.

[the remark this answers is not in the document]
Comment (PlatformLC) [reply cni/pkg/plugin/plugin.go:223]: Modified the previous change a bit regarding current stage: 
Cleanup all ebpf probes while cni terminating (this should introduce traffic escape for upgrade case, we may track it in #43642)

File: cni/pkg/ambient/net_linux.go [hunk 6]
```diff
@@ -193,6 +193,9 @@ func (s *Server) delPodEbpfOnNode(ip string) error {
 		ifIndex = veth.Attrs().Index
 	}
 
+	if force {
+		return s.ebpfServer.RemovePod([]netip.Addr{ipAddr}, uint32(ifIndex))
+	}
```
Comment (bleggett) [inline cni/pkg/ambient/net_linux.go:198]: Mostly a nit - why the use of a separate public method for this versus adding an extra flag to RedirectArgs?

File: cni/pkg/plugin/plugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bleggett) [reply cni/pkg/plugin/plugin.go:223]: That sounds good to me, thank you!

File: cni/pkg/ambient/net_linux.go (hunk 6 above)
Comment (PlatformLC) [reply cni/pkg/ambient/net_linux.go:198]: Hmm, it's  a bit tricky here. Since channel messaging will be terminated with context.Done(channel message mechanism could guarantee the sequence at least), so we need a "force"-like method for cleanup.

File: cni/pkg/plugin/plugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (PlatformLC) [reply cni/pkg/plugin/plugin.go:223]: @howardjohn , any other comment or idea for it?

File: cni/pkg/ambient/net_linux.go (hunk 6 above)
Comment (bleggett) [reply cni/pkg/ambient/net_linux.go:198]: Ah, ok - that's fine, I wasn't sure.

Long term that makes me wonder if the request-based "server" model might not actually be that helpful versus simpler public methods, but that's out of scope here.

**Review by bleggett — APPROVED** [review]
LGTM - sorry for the long back-and-forth

**Comment by PlatformLC** [comment]
> LGTM - sorry for the long back-and-forth

That's ok for me, I think making things better is the same goal for us. 👍

File: cni/pkg/ambient/net.go [hunk 7]
```diff
@@ -296,3 +298,36 @@ func (s *Server) DelPodFromMesh(pod *corev1.Pod) {
 func SetProc(path string, value string) error {
 	return os.WriteFile(path, []byte(value), 0o644)
 }
+
+func (s *Server) cleanStaleIPs(stales sets.Set[string]) {
+	log.Infof("Ambient stale Pod IPs to be cleaned: %s", stales)
+	switch s.redirectMode {
+	case IptablesMode:
```
Comment (howardjohn) [inline cni/pkg/ambient/net.go:305]: why is there no cleanup for iptables mode? don't we need to remove the pod from IPset?

File: cni/pkg/ambient/informers.go [hunk 8]
```diff
@@ -71,6 +79,7 @@ func (s *Server) EnqueueNamespace(o controllers.Object) {
 				Old:   pod,
 				Event: controllers.EventUpdate,
 			})
+			processed.Insert(pod.Status.PodIP)
```
Comment (howardjohn) [inline cni/pkg/ambient/informers.go:82]: probably want .PodIPs?

File: cni/pkg/ambient/net_linux.go (hunk 1 above)
Comment (howardjohn) [reply cni/pkg/ambient/net_linux.go:1192]: My original comment is still not addressed. We only cleanup for ebpf but not iptables. Regardless of all other details, iptables and eBPF should follow the same cleanup

Comment (howardjohn) [reply cni/pkg/ambient/net_linux.go:1192]: Specifically AnnotateUnenrollPod

File: cni/pkg/ambient/server.go [hunk 9]
```diff
@@ -187,6 +189,12 @@ func (s *Server) ReconcileZtunnel() error {
 	s.UpdateConfig()
 	if activePod == nil {
 		log.Infof("active ztunnel updated, no ztunnel running on the node")
+		if s.redirectMode == EbpfMode {
+			if err := s.delZtunnelEbpfOnNode(); err != nil {
```
Comment (howardjohn) [inline cni/pkg/ambient/server.go:193]: why not just move this into cleanupNode? its pretty hard to follow with all these `if EBPF` without them consolidated in just a few places. 

In the future someone will probably call cleanupNode and have broken code?

File: cni/pkg/plugin/plugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (howardjohn) [reply cni/pkg/plugin/plugin.go:223]: I don't really like this tbh, it will result in downtime during onboarding. One of the key value propositions is that enabling ztunnel doesn't break anything.

File: cni/pkg/ambient/net.go (hunk 7 above)
Comment (PlatformLC) [reply cni/pkg/ambient/net.go:305]: will add the removal to make logic consistently

File: cni/pkg/plugin/plugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (PlatformLC) [reply cni/pkg/plugin/plugin.go:223]: Hmm.. if "availability" is prior to security, I'll revert this piece of change for enforcing to incept traffic for this "on-ramp/off-ramp" state

[the remark this answers is not in the document]
Comment (howardjohn) [reply cni/pkg/plugin/plugin.go:223]: Security is important too, thats why I suggested it should only be on initial onboarding. Once its ready once its permanent. We do not have to compromise on either.

**Comment by PlatformLC** [comment]
@howardjohn @bleggett, I've updated this PR to make eBPF and iptables mode has the same behavior. Would you help PTAL?

BTW, for cleanupNode level rules for upgrade/uninstall case, could we add apps/daemonsets AC for istio-cni SA? So that we could determine the exiting event by looking for if `istio-cni` ds is gone away or deleting(for this case, it should be uninstall, otherwise it's just a restart/upgrade case.) How do you think about it?

File: cni/pkg/ambient/net.go [hunk 10]
```diff
@@ -297,3 +302,39 @@ func (s *Server) DelPodFromMesh(pod *corev1.Pod, event controllers.Event) {
...
+func (s *Server) cleanupPodsEbpfOnNode() error {
+	if s.ebpfServer == nil {
+		return fmt.Errorf("uninitialized ebpf server")
+	}
+	for _, ns := range s.namespaces.List(
+		metav1.NamespaceAll, klabels.Set{pconstants.DataplaneMode: pconstants.DataplaneModeAmbient}.AsSelector()) {
+		namespace := ns.GetName()
+		for _, pod := range s.pods.List(namespace, klabels.Everything()) {
+			log.Infof("cleanup Pod %s in %s", pod.Name, namespace)
+			if err := s.delPodEbpfOnNode(pod.Status.PodIP, true); err != nil {
```
Comment (howardjohn) [inline cni/pkg/ambient/net.go:331]: This will be called on all pods -- but we may have some that explicitly opt out with annotation. So we should either filter those, or this needs to be sure to gracefully handle this.

**Review by howardjohn — APPROVED** [review]
